### DISCUSSION TITLE
Questions regarding deployment and re-deployment

### DISCUSSION THREAD

**Discussion by esdcharles** [Azure/ipam#112] [Q&A]
Hi there folks,

I am still struggling to get this solution up and going using an Application Gateway with my resources behind it leveraging both Private Endpoints and separate containers. With respect to the deployment I was hoping someone can answer the following:

1) Is it possible only to target only a part of the solution  i.e. blob storage, cosmos etc. . In my situation my Cosmos Container, Cosmos DB and Blob did not install. I would like to only target the ones that did not install re-running the deployment. Apologies ahead of time if I missed this option in the documentation.

With the above I am using this to deploy: 

./deploy.ps1 -Location "mylocation" -ResourceNames $ResourceNames -UIAppName "myuiname" -EngineAppName "myenginename" -PrivateACR

and have confirmed that $ResourceNames has values for : storageAccountName, cosmosDatabaseName, cosmosContainerName . Strangely cosmosAccountName always deploys without issue?

2) Similar to Terraform, If I rerun the deployment multiple times will the bicep files be skipped should the resources already exist from a previous install? With folks like myself who are struggling with this it would be nice if it did rather than having to delete everything and start over each time I am troubleshooting a single resource. 

Thanks for your assistance!

**Comment by DCMattyG** [comment]
Good morning @binarybfc!

Here are my answers to your questions above:

1. At this time, we don't have any special switches in the Deployment script to partially deploy the solution. There are many components of the deployment which are tightly coupled (e.g. Managed Identity tied to all resources) and covering all of the potential cases could become quite complex. I do agree that this would be a "nice to have", but I think our strategy here will be more around documentation on how to properly re-deploy specific components, rather than packing that into the current script (which is already > 1000 lines long, lol).
2. Bicep is idempotent. Whilst I haven't tested it (and I will), if you re-deploy the exact same resources, in theory those pieces should be skipped. The one "gotcha" I can think of right now is that the App Registrations change with each deploy, and as such some of the child resources will be "different" which would affect some of this child Bicep resources, making them "different" in some ways.

The bottom line is that the Deployment script is intended to be the "easy button", but the trade-off is that it's tightly coupled for the time being. As we move into the future, our intention is to further document the individual components so customers can better self-service deploying a subset of the overall resources, or perhaps re-deploy individual resources on an as-needed basis.

For now, we are **more** than happy to help jump on a call with you to get you up and running. Please do not hesitate to reach out to me at Matthew.Garrett@microsoft.com and we can schedule some time (at your convenience) to sync up live.

++ @hbendana for visibility as well
